File: vigour-app/src/lib/mcp-client.ts
const MCP_URL = "http://localhost:3002";

export interface McpToolResult {
  content: Array<{ type: "text"; text: string }>;
  isError?: boolean;
}

interface McpRequest {
  jsonrpc: "2.0";
  id: number;
  method: string;
  params?: unknown;
}

let requestId = 0;

async function sendRequest(method: string, params?: unknown): Promise<unknown> {
  const id = ++requestId;
  const body: McpRequest = { jsonrpc: "2.0", id, method, params };

  const res = await fetch(MCP_URL, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  });

  if (!res.ok) {
    throw new Error(`MCP request failed: ${res.status} ${res.statusText}`);
  }

  const data = await res.json();
  if (data.error) {
    throw new Error(`MCP error: ${data.error.message ?? JSON.stringify(data.error)}`);
  }

  return data.result;
}

/** List all available tools from the MCP server. */
export async function listTools(): Promise<string[]> {
  const result = await sendRequest("tools/list") as { tools: Array<{ name: string }> };
  return result.tools.map((t) => t.name);
}

/** Call a tool by name with the given arguments. */
export async function callTool(name: string, args: Record<string, unknown>): Promise<McpToolResult> {
  const result = await sendRequest("tools/call", { name, arguments: args }) as McpToolResult;
  return result;
}

File: vigour-app/cortex/vigour_cortex/mcp_client.py
"""MCP HTTP client to invoke vigour-core tools."""

from __future__ import annotations

import json
import logging
from typing import Any

import httpx

logger = logging.getLogger(__name__)


class McpClient:
    """Speaks MCP JSON-RPC over HTTP to vigour-core on :3002."""

    def __init__(self, url: str) -> None:
        self._url = url
        self._req_id = 0
        self._http = httpx.AsyncClient(timeout=30)

    async def close(self) -> None:
        await self._http.aclose()

    async def _request(self, method: str, params: Any = None) -> Any:
        self._req_id += 1
        body = {"jsonrpc": "2.0", "id": self._req_id, "method": method}
        if params is not None:
            body["params"] = params
        res = await self._http.post(self._url, json=body)
        res.raise_for_status()
        data = res.json()
        if "error" in data:
            raise RuntimeError(f"MCP error: {data['error']}")
        return data["result"]

    async def list_tools(self) -> list[str]:
        result = await self._request("tools/list")
        return [t["name"] for t in result["tools"]]

    async def call_tool(self, name: str, args: dict[str, Any]) -> str:
        result = await self._request("tools/call", {"name": name, "arguments": args})
        texts = [c["text"] for c in result["content"] if c["type"] == "text"]
        return "\n".join(texts)

    async def parse_intent(self, text: str) -> str:
        """Ask vigour-core to parse a natural-language intent and execute it."""
        return await self.call_tool("unrecognized", {"originalQuery": text})
